File: sections/14_core_patterns_mixed_contest/problems/m_required_letter_subsequence/solution.py
import sys


def main() -> None:
    data = sys.stdin.buffer.read().split()
    s = data[0].decode()
    k = int(data[1])
    required = data[2].decode()
    r = int(data[3])
    remaining_required = s.count(required)
    chosen_required = 0
    answer = []
    for i, char in enumerate(s):
        while (answer and answer[-1] > char
               and len(answer) - 1 + len(s) - i >= k
               and (answer[-1] != required
                    or chosen_required - 1 + remaining_required >= r)):
            if answer[-1] == required:
                chosen_required -= 1
            answer.pop()
        if len(answer) < k:
            if char == required:
                answer.append(char)
                chosen_required += 1
            elif k - len(answer) - 1 >= r - chosen_required:
                answer.append(char)
        if char == required:
            remaining_required -= 1
    print("".join(answer))


if __name__ == "__main__":
    main()

File: sections/14_core_patterns_mixed_contest/problems/m_required_letter_subsequence/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    string s, answer;
    int k, r;
    char required;
    cin >> s >> k >> required >> r;
    int remaining_required = count(s.begin(), s.end(), required);
    int chosen_required = 0;
    for (int i = 0; i < (int)s.size(); ++i) {
        char c = s[i];
        while (!answer.empty() && answer.back() > c
               && (int)answer.size() - 1 + (int)s.size() - i >= k
               && (answer.back() != required
                   || chosen_required - 1 + remaining_required >= r)) {
            if (answer.back() == required) --chosen_required;
            answer.pop_back();
        }
        if ((int)answer.size() < k) {
            if (c == required) {
                answer.push_back(c);
                ++chosen_required;
            } else if (k - (int)answer.size() - 1 >= r - chosen_required) {
                answer.push_back(c);
            }
        }
        if (c == required) --remaining_required;
    }
    cout << answer << '\n';
}
